### PR TITLE
First phase of high-availability objects.

### PR DESCRIPTION
**Pull request by kevinsteves**
Includes all config for active passive, HA interfaces and link
monitoring groups.

TODO: active active, cluster, path monitoring groups.

### CODE REVIEW

File: dev/ha/config.go [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+	Ha3       *ha3Interface       `xml:"ha3"`
+	Ha4       *ha4Interface       `xml:"ha4"`
+	Ha4Backup *ha4Interface       `xml:"ha4-backup"`
+}
+
+type common struct {
+	Port      string `xml:"port,omitempty"`
+	IpAddress string `xml:"ip-address,omitempty"`
+	Netmask   string `xml:"netmask,omitempty"`
+}
```
Comment (shinmog) [inline dev/ha/config.go:395]: Please ditch this anonymous struct; just put the params in the structs that need them.

**Review by shinmog — CHANGES_REQUESTED** [review]
Nice work!

I noticed that there are no unit tests with either of these.  Unit tests are there to help make sure that your marshaling / unmarshaling and `Copy()` are working properly.  Especially with only one version on each of these namespaces, the unit tests aren't that bad.  You don't have to go ham on unit tests like I've done for some of the namespaces, but I'd like to see a few unit tests for these namespaces as well.  You can use netw/routing/protocol/bgp as the basis for the ha unit tests:  just copy "fw_test.go" over (updating calls to `Get()` and `Set()` as appropriate), and copy "testdata_test.go" over (updating the Config as appropriate).

**Comment by kevinsteves** [comment]
Address changes requested.

**Review by shinmog — APPROVED** [review]
Ok cool :)

The HA object is huge, and it's much easier in practice to have your unittests only do certain chunks of config in each test.  This is because if there is a bug, looking at a huge struct dump comparing what was sent with what was expected makes it very hard to see what the difference is.  But great work on this!

**Pull request merged**